### PR TITLE
JENKINS-36547: Introduce Queue.Task.getAffinityKey()

### PR DESCRIPTION
**Pull request by orgi**
...and use it in `LoadBalancer.assignGreedily()` to find a node for a task.

See [JENKINS-36547](https://issues.jenkins-ci.org/browse/JENKINS-36547).

### Proposed changelog entries

* Entry 1:
    - type: bug
      message: >
        Prepare `Queue` API to be able to provide job-specific affinity key so that JENKINS-36547 can be fixed
      references:
        - issue: 36547

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers
@jglick

### CODE REVIEW

**Comment by jglick** [comment]
Assuming you fix the Javadoc error so this becomes buildable and is deployed to the Incrementals repository, you want to file a matching PR to (I suppose) `workflow-durable-task-step-plugin` which would pick this up via the `jenkins.version` property and use the new API, preferably with a functional test there demonstrating that it actually solves the Pipeline issue.

**Comment by orgi** [comment]
Thanks for the hint about the Incrementals repository. Will try to get it running locally and finally file the pr for workflow-durable-taks-step-plugin as soon as it's running.

File: core/src/main/java/hudson/model/Queue.java [hunk 1]
```diff
@@ -1896,6 +1896,12 @@ default CauseOfBlockage getCauseOfBlockage() {
          */
         String getFullDisplayName();
 
+        /**
+         * Key for this task used for scheduling when searching for the next fee executor
```
Comment (daniel-beck) [inline core/src/main/java/hudson/model/Queue.java]: Typo: fee.

Also, rather difficult to understand. Perhaps rephrase for clarity?

Comment (orgi) [reply core/src/main/java/hudson/model/Queue.java]: Thanks for pointing out the typo. I'll fix that.

How about something like the following for the javadoc:

        /**
         * Returns task-specific key which is used by the (@Link LoadBalancer) to choose one particular executor
         * amongst all the free executors on all possibly suitable nodes.
         * NOTE: To be able to re-use the same node during the next run this key should not change from one run to
         * another. You probably want to compute that key based on the job's name.
         * <p>
         * @return by default: (@Link #getFullDisplayName())
         *
         * @see hudson.model.LoadBalancer
         */

Comment (daniel-beck) [reply core/src/main/java/hudson/model/Queue.java]: LGTM.

File: core/src/main/java/hudson/model/Queue.java [hunk 2]
```diff
@@ -1896,6 +1896,12 @@ default CauseOfBlockage getCauseOfBlockage() {
...
+         * @see hudson.model.LoadBalancer
+         */
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Queue.java]: nice to have `@since TODO`

File: core/src/main/java/hudson/model/Queue.java (hunk 1 above)
Comment (oleg-nenashev) [reply core/src/main/java/hudson/model/Queue.java]: LGTM

**Comment by daniel-beck** [comment]
Ideally has a linked downstream PR showing this in action, like @jglick described.

**Review by oleg-nenashev — COMMENTED** [review]
New documentation proposal is OK, IMHO fine to merge once it is integrated

**Comment by orgi** [comment]
Sorry for the delay about the PR for the workflow-durable-task-step-Plugin. Having a hard time getting things running behind our corp proxy.

**Comment by orgi** [comment]
Just pushed the updated JavaDoc. Wanted to wait until I'm able to use the incrementals build behind our proxy. However I didn't find time so far to take a closer look what's going wrong here.
Just in case you're willing to already pull this without the PR for the durable-task-step-plugin the JavaDoc should be fine now.

The durable-task-step-plugin PR will take me some time as there have been some changes I need to merge. In addition to that I want to add some more tests than initially thought for better test coverage.

**Comment by orgi** [comment]
I have everything prepared for the PR of the `workflow-durable-task-step-plugin`. Unfortunately I could not get the build picking the corresponding Jenkins incrementals by using `mvn incrementals:update`. I could only use them hard coding the incrementals version `2.145-rc27341.296ad9122b93` into the `pom.xml` of `workflow-durable-task-step-plugin`. However I assume it'd be a bad idea to include that one in the PR...
Can you give me any pointers on how to ensure that the plugin maven build will pick the Jenkins incrementals build I need so I can file the PR for the `workflow-durable-task-step-plugin` which will be using the Jenkins build created by this PR?

**Review by oleg-nenashev — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
@orgi https://github.com/jenkinsci/label-verifier-plugin/pull/4 may be a good example. `consume-incrementals` profile on the recent Plugin POM should be enough

**Comment by orgi** [comment]
@oleg-nenashev: Thanks, looks like I did it correct, though. I'll then proceed with the PR. :)

**Comment by orgi** [comment]
I filed the PR (see above). However for some reason it doesn't seem to pick the correct Jenkins incremental build which is configured in the pox.xml. Therefore the build fails.

Locally everything is running just fine (it picks up the incremental build and runs the tests).

Looks like I once more need some help on how to use the incrementals correctly.

**Comment by orgi** [comment]
Fortunately @dwnusbaum got the related PR jenkinsci/workflow-durable-task-step-plugin#79 running as it needed some change in the Jenkinsfile.

@jglick: Ok to merge? Is there still something to be done from my side?

**Review by jglick — APPROVED** [review]

**Pull request merged**